### PR TITLE
3 patches for Git external

### PR DESCRIPTION
**Pull request by biddisco**

### CODE REVIEW

**Comment by biddisco** [comment]
When the initial clone is performed, the script prints out the "git clone http://xxxx options" information, but my addition of the shallow clone option made the print look like "git clonehttp://xxxxoptions" without spaces, so I have added a simply string replace utility to make it look nice again.

**Comment by biddisco** [comment]
rebased onto master to remove a merge commit

File: GitExternal.cmake [hunk 1]
```diff
@@ -8,21 +8,37 @@
...
 #    or in the given absolute path using the given repository and tag
 #    (commit-ish).
 #
-# Options which control behaviour:
+# Options which can be supplied to the function
+#  VERBOSE, when present, this option tells the function to output
+#    information about what operations are being performed by git on 
+#    the repo.
+#  DISABLE_UPDATE, when present, once the repo has been initially cloned and 
+#    setup then no further git operations wil be performed. GitExternal will not
```
Comment (eile) [inline GitExternal.cmake]: typo wil

File: GitExternal.cmake [hunk 2]
```diff
@@ -52,13 +68,20 @@ set(GIT_EXTERNAL_USER_FORK ${GIT_EXTERNAL_USER} CACHE STRING
...
+# utility function for printing a list with custom separator
+function(JOIN VALUES GLUE OUTPUT)
+  string (REGEX REPLACE "([^\\]|^);" "\\1${GLUE}" _TMP_STR "${VALUES}")
+  string (REGEX REPLACE "[\\](.)" "\\1" _TMP_STR "${_TMP_STR}") #fixes escaping
+  set (${OUTPUT} "${_TMP_STR}" PARENT_SCOPE)
+endfunction()
+
 function(GIT_EXTERNAL DIR REPO TAG)
-  cmake_parse_arguments(GIT_EXTERNAL "" "" "RESET" ${ARGN})
+  cmake_parse_arguments(GIT_EXTERNAL_LOCAL_OPTION "DISABLE_UPDATE;VERBOSE;SHALLOW" "" "RESET" ${ARGN})
```
Comment (eile) [inline GitExternal.cmake]: The ..._RESET needs to be renamed below now.

File: GitExternal.cmake [hunk 3]
```diff
@@ -80,15 +103,28 @@ function(GIT_EXTERNAL DIR REPO TAG)
...
   get_filename_component(GIT_EXTERNAL_DIR "${DIR}/.." ABSOLUTE)
 
+  set(FRESH_CLONE 0)
   if(NOT EXISTS "${DIR}")
-    message(STATUS "git clone ${REPO} ${DIR}")
+    set(FRESH_CLONE 1)
+    if (GIT_EXTERNAL_LOCAL_OPTION_SHALLOW)
+      set(_command clone --recursive --depth 1 --branch ${TAG} ${REPO} ${DIR})
+    else()
+      set(_command clone " " --recursive " " ${REPO} " " ${DIR})
```
Comment (eile) [inline GitExternal.cmake]: why the "?

File: GitExternal.cmake [hunk 4]
```diff
@@ -80,15 +103,28 @@ function(GIT_EXTERNAL DIR REPO TAG)
...
+    endif()
+    JOIN("${_command}" " " _msg_text)
+    message(STATUS "git " ${_msg_text})
```
Comment (eile) [inline GitExternal.cmake]: get_external_message

**Comment by biddisco** [comment]
@eile I have fixed the doc typos, extra ", GIT_EXTERNAL_LOCAL_OPTION_RESET and git_external_message and amended the relevant commits.

**Comment by eile** [comment]
We're in a bit of a cmake fire right now, since the massive cleanup from yesterday is still broken. Will look at this PR in a couple of days, when things have settled down

**Comment by biddisco** [comment]
I've merged your big changeset int and fixed the couple of small typos I found. AFAICT this is working ok and can be merged.

**Comment by eile** [comment]
Merged in #366

**Pull request closed without merging**